lib/imap/providers/generic.rb

```diff
@@ -0,0 +1,112 @@
+# frozen_string_literal: true
+
+require 'net/imap'
+
+module Imap
+  module Providers
+    class Generic
+
+      def initialize(server, options = {})
+        @server = server
+        @port = options[:port] || 993
+        @ssl = options[:ssl] || true
+        @username = options[:username]
+        @password = options[:password]
+        @timeout = options[:timeout] || 10
+      end
+
+      def imap
+        @imap ||= Net::IMAP.new(@server, port: @port, ssl: @ssl, open_timeout: @timeout)
+      end
+
+      def connect!
+        imap.login(@username, @password)
+        @capabilities = imap.responses['CAPABILITY'][-1]
+      end
+
+      def disconnect!
+        imap.logout rescue nil
+        imap.disconnect
+      end
+
+      def can?(capability)
+        @capabilities.include?(capability)
+      end
+
+      def uids(opts = {})
+        if opts[:from] && opts[:to]
+          imap.uid_search("UID #{opts[:from]}:#{opts[:to]}")
+        elsif opts[:from]
+          imap.uid_search("UID #{opts[:from]}:*")
+        elsif opts[:to]
+          imap.uid_search("UID 1:#{opts[:to]}")
+        else
+          imap.uid_search('ALL')
+        end
+      end
+
+      def labels
+        @labels ||= begin
+          labels = {}
+
+          list_mailboxes.each do |name|
+            if tag = to_tag(name)
+              labels[tag] = name
+            end
+          end
+
+          labels
+        end
+      end
+
+      def open_mailbox(mailbox_name, write = false)
```

Comment:
I dislike boolean parameters like this, because when you read the code at the calling point it looks like this:

`open_mailbox('my_mailbox', true)` and it's totally unclear what `true` means.

Instead we should use a keyword argument like `open_mailbox('my_mailbox', write: true)` which is self documenting.